### PR TITLE
Subtoolchain searching

### PR DESCRIPTION
**Pull request by ocaisa**
Add support for automatically matching the lowest possible member of a toolchain hierarchy when resolving dependencies using a command line option ```--minimal-toolchains```.
Requires a ```SUBTOOLCHAIN``` to be defined for each level in the hierarchy (see examples inside)

WIP @boegel ... please review
FYI @geimer

### CODE REVIEW

**Comment by boegel** [comment]
Jenkins: ok to test

**Comment by boegel** [comment]
@ocaisa: there's a lot going on here, and I have very little time left for EB v2.2, so realistically speaking I need to postpone this for EB v2.3...

One major aspect to be tackled here is a couple of unit tests, and probably some code cleanup.

Let's rediscuss this in August?

**Comment by ocaisa** [comment]
Yeah, I know its a bit of a monster...August it is, I'll give it a field test before then, unit tests be damned!

On 7 Jul 2015 8:00 pm, "Kenneth Hoste" <notifications@github.com<mailto:notifications@github.com>> wrote:

@ocaisa<https://github.com/ocaisa>: there's a lot going on here, and I have very little time left for EB v2.2, so realistically speaking I need to postpone this for EB v2.3...

One major aspect to be tackled here is a couple of unit tests, and probably some code cleanup.

Let's rediscuss this in August?

—
Reply to this email directly or view it on GitHub<https://github.com/hpcugent/easybuild-framework/pull/1306#issuecomment-119284788>.


------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------
Forschungszentrum Juelich GmbH
52425 Juelich
Sitz der Gesellschaft: Juelich
Eingetragen im Handelsregister des Amtsgerichts Dueren Nr. HR B 3498
Vorsitzender des Aufsichtsrats: MinDir Dr. Karl Eugen Huthmacher
Geschaeftsfuehrung: Prof. Dr.-Ing. Wolfgang Marquardt (Vorsitzender),
Karsten Beneke (stellv. Vorsitzender), Prof. Dr.-Ing. Harald Bolt,
Prof. Dr. Sebastian M. Schmidt
------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------

**Comment by boegel** [comment]
@ocaisa: last tests have failed due to an issue on Jenkins, I'm looking into it

**Comment by boegel** [comment]
Jenkins: test this please

**Comment by ocaisa** [comment]
@boegel Is this a Jenkins problem?

**Comment by boegel** [comment]
@ocaisa: no, you broke something it seems

```
ERROR [35.062s]: Test --read-only-installdir and --group-write-installdir.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test/framework/toy_build.py", line 503, in test_toy_permissions_installdir
    self.test_toy_build(extra_args=['--group-writable-installdir'])
  File "test/framework/toy_build.py", line 146, in test_toy_build
    self.check_toy(self.test_installpath, outtxt, versionsuffix=versionsuffix)
  File "test/framework/toy_build.py", line 80, in check_toy
    self.assertTrue(success.search(outtxt), "COMPLETED message found in '%s" % outtxt)
TypeError: expected string or buffer
```

Let's retrigger the test to rule out a fluke on Jenkins.

**Comment by boegel** [comment]
Jenkins: test this please

**Comment by ocaisa** [comment]
Yup, fluke

**Comment by boegel** [comment]
@ocaisa: gave this a quick look, looks fairly OK at first sight, although the extra code added to `easyconfig/tools.py` looks very verbose, may benefit from some refactoring

I do have some concerns with some changes left and right, of which I'm not sure if they're covered by the tests (in the sense that, I'm not sure if some of the changes will not impact other stuff and go unnoticed with the current tests).

As you know, I'm very keen on getting this included for EasyBuild v2.4.0, and it seems like it's doable given a little bit more work (mostly w.r.t. unit tests & some code cleanup).

Would you have time next Friday for a one-to-one conf call on this to step through the whole diff together?

**Comment by ocaisa** [comment]
@boegel By next Friday do you mean tomorrow or next week (during your vacation!)? I can do a few hours tomorrow and next week I can do Friday afternoon only.

**Comment by boegel** [comment]
@ocaisa: Tomorrow (Oct 29th), let's sync up via IM :)

File: easybuild/framework/easyconfig/easyconfig.py [hunk 1]
```diff
@@ -722,6 +722,26 @@ def get(self, key, default=None):
         else:
             return default
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: `__ne__` should also be implemented, see https://github.com/ocaisa/easybuild-framework/pull/17

File: easybuild/framework/easyconfig/easyconfig.py [hunk 2]
```diff
@@ -722,6 +722,26 @@ def get(self, key, default=None):
...
+            """Make a hashable value of the given value."""
+            if isinstance(val, list):
+                val = tuple([make_hashable(x) for x in val])
+            elif isinstance(val, dict):
+                val = tuple([(key, make_hashable(val)) for (key, val) in sorted(val.items())])
+            return val
+
+        tup = ()
+        for (key, val) in sorted(self.asdict().items()):
+            tup += (key, make_hashable(val))
```
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: this is constructing quite a bit of tuples internally (i.e., it's slow), and there's probably no need to...

we should just use a list here?

```python
hashable = []
for (key, val) in sorted(self.asdict().items()):
    hashable.extend([key, make_hashable(val)])

return hash(hashable)
```

if the tests I added in https://github.com/ocaisa/easybuild-framework/pull/17 still pass with that in place, it's OK

File: easybuild/framework/easyconfig/tools.py [hunk 3]
```diff
@@ -140,6 +141,225 @@ def find_resolved_modules(unprocessed, avail_modules, retain_all_deps=False):
 
     return ordered_ecs, new_unprocessed, new_avail_modules
 
+def toolchain_hierarchy_cache(func):
+    mydict = dict()
```
Comment (boegel) [inline easybuild/framework/easyconfig/tools.py]: missing docstring, and use `cache = {}` here

where did you get inspiration for this BTW?

I've never seen a cache implemented via a decorator function before... :-)

File: easybuild/framework/easyconfig/tools.py [hunk 4]
```diff
@@ -140,6 +141,225 @@ def find_resolved_modules(unprocessed, avail_modules, retain_all_deps=False):
...
+            return mydict[myarg]
+        else:
+            new_val = func(arg)
+            mydict[myarg] = new_val
+            return new_val
+    return wrapped_func
+
+@toolchain_hierarchy_cache
+def get_toolchain_hierarchy(parent_toolchain):
+    # Grab all possible subtoolchains
```
Comment (boegel) [inline easybuild/framework/easyconfig/tools.py]: missing docstring

File: easybuild/framework/easyconfig/tools.py [hunk 5]
```diff
@@ -140,6 +141,225 @@ def find_resolved_modules(unprocessed, avail_modules, retain_all_deps=False):
...
+    _, all_tc_classes = search_toolchain('')
+    subtoolchains = dict((tc_class.NAME, getattr(tc_class, 'SUBTOOLCHAIN', None)) for tc_class in all_tc_classes)
+    # The parent is the first element in the list
+    toolchain_list = [parent_toolchain]
+    current = parent_toolchain
+    while True:
```
Comment (boegel) [inline easybuild/framework/easyconfig/tools.py]: bleh, can we think of a better condition?

from the looks of it, how about, just collapse things "one level up", like this?

```python
while subtoolchains[current['name']]:
```

File: easybuild/framework/easyconfig/tools.py [hunk 6]
```diff
@@ -140,6 +141,225 @@ def find_resolved_modules(unprocessed, avail_modules, retain_all_deps=False):
...
+            _log.debug("Using cache to return hierarchy for toolchain %s" % arg)
+            return mydict[myarg]
+        else:
+            new_val = func(arg)
+            mydict[myarg] = new_val
+            return new_val
+    return wrapped_func
+
+@toolchain_hierarchy_cache
+def get_toolchain_hierarchy(parent_toolchain):
```
Comment (boegel) [inline easybuild/framework/easyconfig/tools.py]: this function is begging for a good unit test, possibly covering all existing toolchains (just to be thorough)...

Comment (boegel) [inline easybuild/framework/easyconfig/tools.py]: ah, wait, the latter would imply having easyconfigs around for all of them, so maybe not ;)

File: easybuild/toolchains/gcc.py [hunk 7]
```diff
@@ -29,8 +29,10 @@
 """
 
 from easybuild.toolchains.compiler.gcc import Gcc
+from easybuild.tools.toolchain import DUMMY_TOOLCHAIN_NAME
 
 
 class GccToolchain(Gcc):
     """Simple toolchain with just the GCC compilers."""
     NAME = 'GCC'
+    SUBTOOLCHAIN = DUMMY_TOOLCHAIN_NAME
```
Comment (boegel) [inline easybuild/toolchains/gcc.py]: this should be implied imho, or maybe the default in `Toolchain`?

File: easybuild/framework/easyconfig/tools.py [hunk 8]
```diff
@@ -140,6 +141,225 @@ def find_resolved_modules(unprocessed, avail_modules, retain_all_deps=False):
...
+                    toolchain_list += [dep_tcs[0]]
+            elif len(unique_versions) == 0:
+                # Check if we have dummy toolchain
+                if subtoolchains[current['name']] == DUMMY_TOOLCHAIN_NAME:
+                    if build_option('add_dummy_to_minimal_toolchains'):
+                        toolchain_list += [{'name': DUMMY_TOOLCHAIN_NAME, 'version': ''}]
+                    break
+                else:
+                    _log.info("Your toolchain hierarchy is not fully populated!")
+                    EasyBuildError("No version found for subtoolchain %s in dependencies of %s"
```
Comment (boegel) [inline easybuild/framework/easyconfig/tools.py]: missing `raise`

File: easybuild/framework/easyconfig/easyconfig.py (hunk 2 above)
Comment (boegel) [inline easybuild/framework/easyconfig/easyconfig.py]: Fixed this myself as a part of https://github.com/ocaisa/easybuild-framework/pull/17

File: easybuild/framework/easyconfig/format/format.py [hunk 9]
```diff
@@ -53,7 +53,7 @@
 DEPENDENCY_PARAMETERS = ['builddependencies', 'dependencies', 'hiddendependencies']
 
 # values for these keys will not be templated in dump()
-EXCLUDED_KEYS_REPLACE_TEMPLATES = ['description', 'easyblock', 'homepage', 'name', 'toolchain', 'version']
+EXCLUDED_KEYS_REPLACE_TEMPLATES = ['description', 'easyblock', 'homepage', 'name', 'toolchain', 'version'] + DEPENDENCY_PARAMETERS
```
Comment (boegel) [inline easybuild/framework/easyconfig/format/format.py]: line too long

**Comment by boegel** [comment]
@ocaisa: fixed broken test, and refactored the `find_resolved_modules` and `find_minimal_resolved_modules` functions in https://github.com/ocaisa/easybuild-framework/pull/21; I verified that it still works as it should (see examples below)

```
$ eb Python-2.7.10-foss-2015b.eb -D | grep zlib 
 * [ ] $CFGS/z/zlib/zlib-1.2.8.eb (module: zlib/1.2.8)
 * [ ] $CFGS/z/zlib/zlib-1.2.8-GCC-4.9.3-binutils-2.25.eb (module: zlib/1.2.8-GCC-4.9.3-binutils-2.25)
 * [ ] $CFGS/z/zlib/zlib-1.2.8-foss-2015b.eb (module: zlib/1.2.8-foss-2015b)
```

```
$ eb Python-2.7.10-foss-2015b.eb -D --minimal-toolchains | grep zlib 
 * [ ] $CFGS/minimal-easyconfigs-0v1IpE/zlib-1.2.8.eb (module: zlib/1.2.8)
 * [ ] $CFGS/minimal-easyconfigs-glxlOb/zlib-1.2.8-GCC-4.9.3-binutils-2.25.eb (module: zlib/1.2.8-GCC-4.9.3-binutils-2.25)
```

```
$ eb Python-2.7.10-foss-2015b.eb -D --minimal-toolchains --add-dummy-to-minimal-toolchains | grep zlib 
 * [ ] $CFGS/minimal-easyconfigs-ry5bSC/zlib-1.2.8.eb (module: zlib/1.2.8)
```

open questions:

* do we always need to dump all easyconfigs involved, or can we detect which ones were fiddled with (i.e., where we changed the toolchain of the dependencies) and only dump those?

TODO:

* [ ] unit test(s) for:
    * [ ] `find_resolved_modules` (already there?)
    * [ ] `find_minimal_resolved_modules`
    * [ ] `robot_find_minimal_easyconfig_for_dependency`
    * new options for `resolve_dependencies`
    * new configure options
* [ ] figure out where we should change `--add-dummy-to-minimal-toolchains` to something like `--consider-dummy-first-for-minimal-toolchains` (needs a good motivating example)
   * basically, sometimes not considering `dummy` will not help us in getting rid of hardcoding dependencies to `dummy`, I think
* [ ] documentation (probably a separate page explaining the differences)

**Comment by ocaisa** [comment]
Not to mix everything in the same bag BUT if you had #1451 then the only time you would be using `dummy` was when you explicitly had to.
As regards only dumping modified easyconfigs, that is pretty easy but would require reparsing the old_spec and comparing it to what you're about to dump. I actually like dumping everything, I think we can manipulate that when it comes to working on reproducability.

**Comment by boegel** [comment]
@ocaisa: how would you feel finishing this, but slapping an 'experimental' tag on it (i.e., require the use of `--experimental`), for now?

I have the feeling we don't have a clear view on the possible issues that may pop up with this...

**Comment by ocaisa** [comment]
I'm OK with that.

On 30 Oct 2015 22:52, "Kenneth Hoste" <notifications@github.com<mailto:notifications@github.com>> wrote:

@ocaisa<https://github.com/ocaisa>: how would you feel finishing this, but slapping an 'experimental' tag on it (i.e., require the use of --experimental), for now?

I have the feeling we don't have a clear view on the possible issues that may pop up with this...

—
Reply to this email directly or view it on GitHub<https://github.com/hpcugent/easybuild-framework/pull/1306#issuecomment-152659409>.


------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------
Forschungszentrum Juelich GmbH
52425 Juelich
Sitz der Gesellschaft: Juelich
Eingetragen im Handelsregister des Amtsgerichts Dueren Nr. HR B 3498
Vorsitzender des Aufsichtsrats: MinDir Dr. Karl Eugen Huthmacher
Geschaeftsfuehrung: Prof. Dr.-Ing. Wolfgang Marquardt (Vorsitzender),
Karsten Beneke (stellv. Vorsitzender), Prof. Dr.-Ing. Harald Bolt,
Prof. Dr. Sebastian M. Schmidt
------------------------------------------------------------------------------------------------
------------------------------------------------------------------------------------------------

**Comment by boegel** [comment]
@ocaisa: OK, I'll try to look into tackling the remainder of the tests.

**Comment by boegel** [comment]
@ocaisa: with https://github.com/ocaisa/easybuild-framework/pull/22 included, I think this is good to go in (but I need to take another close look once that PR is included here)

I also wrote up some documentation for this, see https://github.com/ocaisa/easybuild-framework/pull/22 .

**Comment by boegel** [comment]
@ocaisa: also, there are some minor conflicts in `easybuild/toolchains/ipsmpi.py` and `easybuild/toolchains/gpsmpi.py`, can you look into updating your branch with current `develop` and fixing them?

File: easybuild/toolchains/gcc.py (hunk 7 above)
Comment (boegel) [inline easybuild/toolchains/gcc.py]: nvm, more explicit this way

File: easybuild/tools/options.py [hunk 10]
```diff
@@ -244,6 +246,8 @@ def override_options(self):
                       None, 'store', None),
             'update-modules-tool-cache': ("Update modules tool cache file(s) after generating module file",
                                           None, 'store_true', False),
+            'use-existing_modules': ("Use existing modules when resolving dependencies with minimal toolchains",
```
Comment (boegel) [inline easybuild/tools/options.py]: crap, typo here, `_` should be `-`

**Comment by boegel** [comment]
Going in after extensive testing by rebuilding all HPL easyconfigs, right in time for EasyBuild v2.4.0.

Thanks @ocaisa!
